core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java

```diff
@@ -533,4 +575,121 @@ private static void assertExceptionAsJson(Exception e, String expectedJson) thro
+                expected.addMetadata("es.bytes_wanted", "123");
+                expected.addMetadata("es.bytes_limit", "456");
+                break;
+            case 2:
+                actual = new SearchParseException(new TestSearchContext(null), "Parse failure", new XContentLocation(12, 98));
+                expected = new ElasticsearchException("Elasticsearch exception [type=search_parse_exception, reason=Parse failure]");
+                expected.addMetadata("es.line", "12");
+                expected.addMetadata("es.col", "98");
+                break;
+            case 3:
+                actual = new IllegalArgumentException("Closed resource", new RuntimeException("Resource"));
+                expected = new ElasticsearchException("Elasticsearch exception [type=illegal_argument_exception, reason=Closed resource]",
+                                new ElasticsearchException("Elasticsearch exception [type=runtime_exception, reason=Resource]"));
+                break;
+            case 4:
+                actual = new SearchPhaseExecutionException("search", "all shards failed",
+                            new ShardSearchFailure[]{
+                                    new ShardSearchFailure(new ParsingException(1, 2, "foobar", null),
+                                            new SearchShardTarget("node_1", new Index("foo", "_na_"), 1))
+                            });
+                expected = new ElasticsearchException("Elasticsearch exception [type=search_phase_execution_exception, " +
+                        "reason=all shards failed]");
+                expected.addMetadata("es.grouped", "true");
+                expected.addMetadata("es.phase", "search");
+                break;
+            case 5:
+                actual = new ElasticsearchException("Parsing failed",
+                            new ParsingException(9, 42, "Wrong state",
+                                new NullPointerException("Unexpected null value")));
+
+                ElasticsearchException expectedCause = new ElasticsearchException("Elasticsearch exception [type=parsing_exception, " +
+                        "reason=Wrong state]", new ElasticsearchException("Elasticsearch exception [type=null_pointer_exception, " +
+                        "reason=Unexpected null value]"));
+                expectedCause.addMetadata("es.line", "9");
+                expectedCause.addMetadata("es.col", "42");
+
+                expected = new ElasticsearchException("Elasticsearch exception [type=exception, reason=Parsing failed]", expectedCause);
+                break;
+            default:
+                throw new UnsupportedOperationException("No randomized exceptions generated for type [" + type + "]");
+        }
+
+        if (actual instanceof ElasticsearchException) {
+            if (randomBoolean()) {
+                Map<String, List<String>> randomHeaders = new HashMap<>();
+
+                int nbHeaders = randomIntBetween(1, 5);
```

Comment:
super minor nit: in general, when creating maps and lists and knowing the size you could pass it in.